src/common/type/pack.h

```diff
@@ -0,0 +1,505 @@
+
+The overall idea is similar to JSON but IDs are used instead of names, typing is more granular, and the representation is far more
+compact. A pack can readily be converted to JSON but the reverse is not as precise due to loose typing in JSON. A pack is a stream
+format, i.e. it is intended to be read in order from beginning to end.
+
+Fields in a pack are identified by IDs. A field ID is stored as a delta from the previous ID, which is very efficient, but means
+that reading from the middle is generally not practical. The size of the gap between field IDs is important -- a gap of 1 never
+incurs extra cost, but depending on the field type larger gaps may require additional bytes to store the field ID delta.
+
+NULLs are not stored in a pack and are therefore not typed. A NULL is essentially just a gap in the field IDs. Fields that are
+frequently NULL are best stored at the end of an object. When using .defaultNull in write functions a NULL will be written (by
+making a gap in the IDs) if the value matches the default. When using .defaultNull in read functions the default will be returned
+when the field is NULL (i.e. missing). The standard default is the C default for that type (e.g. bool = false, int = 0) but can be
+changed with the .defaultValue parameter.
+
+A pack is an object by default. Objects can store fields, objects, or arrays. Objects and arrays will be referred to collectively as
+containers. Fields contain data to be stored, e.g. integers, strings, etc.
+
+Here is a simple example of a pack:
+
+PackWrite *write = pckWriteNew(buffer);
+pckWriteU64P(write, 77);
+pckWriteBoolP(write, false, .defaultNull = true);
+pckWriteI32P(write, -1, .defaultNull = true, .defaultValue = -1);
+pckWriteStringP(write, STRDEF("sample"));
+pckWriteEndP();
+
+A string representation of this pack is `1:uint64:77,4:str:sample`. The boolean was not stored because we requested NULL on default
+and the default was false (because it was not set to anything else). The int32 field was also not stored but we set an explicit
+default value. Note that there is a gap in the ID stream, which represents the NULL values.
+
+This pack can be read with:
+
+PackRead *read = pckReadNew(buffer);
+pckReadU64P(read);
+pckReadBoolP(read, .defaultNull = true);
+pckReadI32P(read, .defaultNull = true, .defaultValue = -1);
+pckReadStringP(read);
+pckReadEndP();
+
+Note that defaults are not stored in the pack so any defaults that were applied when writing (by setting .defaultNull and
+optionally .defaultValue) must be applied again when reading (by setting .defaultNull and optionally .defaultValue).
+
+If we don't care about the NULLs, another way to read is:
+
+PackRead *read = pckReadNew(buffer);
+pckReadU64P(read);
+pckReadStringP(read, .id = 4);
+pckReadEndP();
+
+This skips the NULLs but in practice it is better to read all the fields just as they were written so constant IDs are not required.
+By default each read and write advances the ID by one.
```

Comment:
What do you think about clarifying this last statement:
```
By default each read and write advances the ID by one, meaning when specifying an ID
to read/write if the ID was already read/written or skipped, an error will occur because it
cannot be read/written again.
```

Reply:
That's a bit awkward. How about:
```
By default each read/write advances the ID by one. If the ID is specified it must be unique and increasing. 
An error will occur if an ID is read/written and then the same ID is read/written again.
```

Reply:
We're missing the "skipping" part. How about: 
```
By default each read/write advances the field ID by one. If an ID is specified it must be unique and increasing, because it will 
advance the field ID beyond the value specified. An error will occur if an ID is attempted to be read/written but the field ID has 
advanced beyond it.
```

Reply:
OK.

Reply:
Clarified in commit 0ea578426aab7a582cb8e1bf2b4cfda5205b0d94